server/src/main/java/org/apache/druid/server/coordinator/CuratorLoadQueuePeonV2.java

```diff
@@ -0,0 +1,418 @@
+  {
+    return failedAssignCount.getAndSet(0);
+  }
+
+  @Override
+  public int getNumberOfSegmentsInQueue()
+  {
+    return segmentsToLoad.size();
+  }
+
+  @Override
+  public void loadSegment(
+      final DataSegment segment,
+      final LoadPeonCallback callback
+  )
+  {
+
+    final SegmentHolder existingHolder = segmentsToLoad.get(segment);
+    if (existingHolder != null) {
+      if ((callback != null)) {
+        existingHolder.addCallback(callback);
+      }
+      return;
+    }
+    log.info("Asking server peon[%s] to load segment[%s]", basePath, segment.getId());
+    queuedSize.addAndGet(segment.getSize());
+    SegmentHolder holder = new SegmentHolder(segment, LOAD, Collections.singletonList(callback));
+    segmentsToLoad.put(segment, holder);
+    processingExecutor.submit((Callable<Void>) () -> {
+      processSegmentChangeRequest(holder);
+      return null;
+    });
+  }
+
+  @Override
+  public void dropSegment(
+      final DataSegment segment,
+      final LoadPeonCallback callback
+  )
+  {
+    final SegmentHolder existingHolder = segmentsToDrop.get(segment);
+    if (existingHolder != null) {
+      if (callback != null) {
+        existingHolder.addCallback(callback);
+      }
+      return;
+    }
+    log.info("Asking server peon[%s] to drop segment[%s]", basePath, segment.getId());
+    SegmentHolder segmentHolder = new SegmentHolder(segment, DROP, Collections.singletonList(callback));
+    segmentsToDrop.put(segment, segmentHolder);
+    processingExecutor.submit((Callable<Void>) () -> {
+      processSegmentChangeRequest(segmentHolder);
+      return null;
+    });
+  }
+
+  @Override
+  public void markSegmentToDrop(DataSegment dataSegment)
+  {
+    segmentsMarkedToDrop.add(dataSegment);
+  }
+
+  @Override
+  public void unmarkSegmentToDrop(DataSegment dataSegment)
+  {
+    segmentsMarkedToDrop.remove(dataSegment);
+  }
+
+  private void processSegmentChangeRequest(SegmentHolder segmentHolder)
+  {
+    try {
+      // Wait for a random duration since we do not want zookeeper getting pounded with create node requests
+      TimeUnit.MILLISECONDS.sleep(ThreadLocalRandom.current().nextInt(MAX_TASK_WAIT));
+      log.info("Server[%s] %s [%s]", basePath, segmentHolder.getType() == LOAD ?
+                                               "loading" : "dropping", segmentHolder.getSegmentId().toString());
+      final String path = ZKPaths.makePath(basePath, segmentHolder.getSegmentId().toString());
+      final byte[] payload = jsonMapper.writeValueAsBytes(segmentHolder.getChangeRequest());
+      curator.create().withMode(CreateMode.EPHEMERAL).forPath(path, payload);
+
+      checkNodeRemovedExecutor.schedule(
+          () -> {
+            try {
+              if (curator.checkExists().forPath(path) != null) {
+                failAssign(segmentHolder, new ISE("%s was never removed! Failing this operation!", path));
+              } else {
+                log.info("%s detected to be removed. ", path);
```

Comment:
This could stand to be at `log.debug`, I think. (In the spirit of #6741 & trying to keep log chatter on the coordinator useful by not being too noisy.)